app/uk/gov/hmrc/personaldetailsvalidation/AssociationMongoRepository.scala

```diff
@@ -0,0 +1,104 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.personaldetailsvalidation
+
+import com.google.inject.ImplementedBy
+
+import org.mongodb.scala.bson.conversions.Bson
+import org.mongodb.scala.model.Indexes.ascending
+import org.mongodb.scala.model.{Filters, FindOneAndUpdateOptions, IndexModel, IndexOptions, Updates}
+import org.mongodb.scala.model.Updates.combine
+import play.api.Logging
+import uk.gov.hmrc.mongo.MongoComponent
+import uk.gov.hmrc.mongo.play.json.PlayMongoRepository
+import uk.gov.hmrc.personaldetailsvalidation.formats.AssociationFormat.associationFormat
+import uk.gov.hmrc.personaldetailsvalidation.model.Association
+
+import java.util.concurrent.TimeUnit
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+trait AssociationRepositoryConstants {
+  val credentialId: String = "credentialId"
+  val sessionId : String = "sessionId"
+  val validationId: String = "validationId"
+  val lastUpdated: String = "lastUpdated"
+}
+
+@ImplementedBy(classOf[AssociationMongoRepository])
+trait AssociationRepository extends AssociationRepositoryConstants {
+  def create(association: Association): Future[Unit]
+  def find(credentialId: String, sessionId: String): Future[Option[Association]]
+}
+
+@Singleton
+class AssociationMongoRepository @Inject() (config: PersonalDetailsValidationMongoRepositoryConfig,
+                                            mongoComponent: MongoComponent)(implicit executionContext: ExecutionContext)
+  extends PlayMongoRepository[Association](
+    collectionName = "association",
+    mongoComponent = mongoComponent,
+    domainFormat = associationFormat,
+    indexes = Seq(
+      IndexModel(
+        ascending("lastUpdated"),
+        indexOptions = IndexOptions().name("expireAfterSeconds").expireAfter(config.collectionTtl.getSeconds, TimeUnit.SECONDS)
+      ),
+      IndexModel(
+        ascending("credentialId", "sessionId"),
+        indexOptions = IndexOptions().name("credentialId_sessionId_index").unique(true)
+      )
+    )
+  ) with AssociationRepository with Logging
+{
+
+  def create(association: Association): Future[Unit] = {
+
+    collection.findOneAndUpdate(
+      filter(association.credentialId, association.sessionId),
+      combine(
+        Updates.set(credentialId, association.credentialId),
+        Updates.set(sessionId, association.sessionId),
+        Updates.set(validationId, association.validationId),
+        Updates.set(lastUpdated, association.lastUpdated)
+      ),
+      FindOneAndUpdateOptions().upsert(true)
+    ).toFuture()
+      .map(_ => ())
```

Comment:
how do we know if this fails, does it throw an exception, should we recover log and then throw the exception?. perhaps with the validation ID in the log only

Reply:
I've added some error handling as above.